### DISCUSSION TITLE
How does one use the MemoryCacheExtensions class?

### DISCUSSION THREAD

**Discussion by JohnD832** [serenity-is/Serenity#6813] [Help] [answered]
I have a method on one of my `StaffRepository.cs` that gets called a lot. I wish to store the value of the 1st call into an `IMemoryCache`.

I tried to do this by passing an `IMemoryCache` via the constructor like how we usually do things in .NET Core.

```
public class StaffRepository
{
	private readonly IMemoryCache _memoryCache;

	public StaffRepository(IMemoryCache memoryCache)
	{
		_memoryCache = memoryCache;
	}
	
	// all other methods
}
```

But this is an existing old code base, and that repository is used a lot in many places like this:

```
var staff = new StaffRepository().GetByUsername(userName);
```

Obviously that will no longer work now cause the constructor requires an IMemoryCache.

After some Googling, I found out that there is `MemoryCacheExtensions` class in Serenity. How do I use that to cache the Database result of `GetByUsername(userName)`?

**Comment by VictorTomaili** [comment]
To use extensions, you need the actual memory cache object. See how [extensions](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods) work.

Repository style implementation had this type of issue. When you need to inject any object into your class, you need to pass it to each call. To avoid that, we removed the repository and injected handlers on endpoints with interface. With that way, constructor injection works automatically. 

But for your case, you can use the `this.Cache.Memory` object directly to access memory cache in Repository and Handlers.

**Comment by JohnD832** [comment]
Apologies @VictorTomaili, are you saying I should have a `this.Cache.Memory` on my `XYZRepository.cs`? Cause right I get the error:

> 'XYZRepository' does not contain a definition for 'Cache' and no accessible extension method 'Cache' accepting a first argument of type 'XYZRepository' could be found (are you missing a using directive or an assembly reference?)

**Comment by VictorTomaili** [comment, marked as answer]
Ensure about your base class. 

https://github.com/serenity-is/Serenity/blob/da44cf06b8f00087647321e5354ad2139b8e913f/src/Serenity.Net.Services/RequestHandlers/Repository/BaseRepository.cs#L22
